File: LeetCode/cpp/106.cpp
class Solution {
public:
    vector<int> inorder;
    vector<int> postorder;

    TreeNode *buildTree(vector<int> &inorder, vector<int> &postorder) {
        int size = inorder.size();
        if (size == 0)
            return nullptr;
        this->inorder = inorder;
        this->postorder = postorder;
        return FindSubtree(0, size - 1, 0, size - 1);
    }

    TreeNode *FindSubtree(int &&i, int &&j, int &&m, int &&n) {
        if (i > j)
            return nullptr;
        TreeNode *node = new TreeNode(postorder[n]);
        int k = i;
        while (k < j) {
            if (inorder[k] == postorder[n])
                break;
            ++k;
        }
        node->left = FindSubtree(i + 0, k - 1, m + 0, m + k - i - 1);
        node->right = FindSubtree(k + 1, j + 0, m + k - i, n - 1);
        return node;
    }
};

File: LeetCode/python3/106.py
class Solution:
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """

        def BuildSub(i, j, m, n):
            if i > j or m > n:
                return None
            node = TreeNode(postorder[n])
            k = inorder.index(postorder[n])
            node.left = BuildSub(i, k-1, m, m + k-i-1)
            node.right = BuildSub(k+1, j, m+k-i, n-1)
            return node

        size = len(inorder)
        if not size:
            return None
        return BuildSub(0, size - 1, 0, size-1)
